### PR TITLE
Add h5py support to NDArrayIter

### PR DESCRIPTION
**Pull request by leezu**
This pull request adds basic h5py support to NDArrayIter.

Data will be read from the h5 dataset on demand (when creating a batch) so that large datasets that don't fit in memory are supported. Shuffling is also supported by storing a set of shuffled indices.

### CODE REVIEW

File: python/mxnet/io.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import ctypes
 import logging
 import threading
+import h5py
```
Comment (piiswrong) [inline python/mxnet/io.py]: we don't want to depend on h5py by default. Do it as an optional import. See cv2 dependency in recordio

Comment (leezu) [inline python/mxnet/io.py]: I changed it. On a side-note, how to specify h5py as a dependency for the tests?

Comment (leezu) [inline python/mxnet/io.py]: @piiswrong

**Comment by Dranithix** [comment]
Rather than establishing a direct dependency on `h5py`, it would be better to attempt to load the library and thereafter create checks as to whether or not loading from h5 is supported on the system no?

**Comment by leezu** [comment]
The dependency is only for the tests. No dependency is added in the python frontend. I haven't figured out yet though how to properly declare it so that h5py actually gets installed on the CI system.

**Comment by leezu** [comment]
@piiswrong anything that should be addressed before merging this?

**Pull request merged**